### ISSUE TITLE
CA1508: false positive on sub-sequential more strict condition

### ISSUE THREAD

**Issue by MaceWindu** [dotnet/roslyn-analyzers#6868]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508): `Avoid dead conditional code`

### Analyzer source

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0-preview.23364.2 (Latest)

### Describe the bug

When variable checked to have one of two values and next checked to have one of those values specifically, CA1508 generated.
For specific code it is possible to workaround (with less code as a bonus) by inverting condition to check value doesn't have remaining possible values.

### Steps To Reproduce

```cs
public class Test
{
	public static void Method(string parameter)
	{
		var flag = GetFlag();

		// triggers CA1508 below
		if (flag == null || flag == false)
		// workaround: works
		//if (flag != true)
		{
			// CA1508  'flag == false' is always 'false'.Remove or refactor the condition(s) to avoid dead code.
			if (flag == false)
			{
			}
		}
	}

	public static bool? GetFlag() => false;
}

```

### Expected behavior

No errors reported

### Actual behavior

Error reported

**Comment by Youssef1313** [comment]
@mavasani I have been investigating this for a while today.

We **incorrectly** hit this condition:

![image](https://github.com/dotnet/roslyn-analyzers/assets/31348972/19db7da1-5f66-485e-baea-7ffdfe9ddd3a)

So we are not merging the results of `flag` being null and false.

This is because the equality implementation relies on hash codes. The hash code of `false` is zero, and we also consider null to be zero.
